python/dgl/sampling/pinsage.py

```diff
@@ -0,0 +1,214 @@
+from .. import function as fn
+from .randomwalks import random_walk
+from .neighbor import select_topk
+from ..base import EID
+
+
+class GenericPinSAGESampler(object):
+    """PinSAGE-like sampler extended to any heterographs, given a metapath.
+
+    Given a heterogeneous graph, PinSAGE neighbor sampler would generate a homogeneous
+    graph where the neighbors of each node are the most commonly visited nodes of the
+    same type by random walk with restarts.  The random walk with restarts are based
+    on a given metapath, which should have the same beginning and ending node type.
+
+    Parameters
+    ----------
+    G : DGLHeteroGraph
+        The bidirectional bipartite graph.
+
+        The graph should only have two node types: ``ntype`` and ``other_type``.
+        The graph should only have two edge types, one connecting from ``ntype`` to
+        ``other_type``, and another connecting from ``other_type`` to ``ntype``.
+
+        PinSAGE works on a bidirectional bipartite graph where for each edge
+        going from node u to node v, there exists an edge going from node v to node u.
+    ntype : str
+        The node type for which the graph would be constructed on.
+    metapath : list[str]
+        The metapath.
+    random_walk_length : int
+        The maximum number of steps of random walk with restarts.
+
+        Note that here we consider traversing from ``ntype`` to ``other_type`` then back
+        to ``ntype`` as a single step (i.e. a single step consists of two hops).
+
+        Usually considered a hyperparameter.
+    random_walk_restart_prob : int
+        Restart probability of random walk with restarts.
+
+        Note that the random walks only would halt on node type ``ntype``, and would
+        never halt on ``other_type``.
+
+        Usually considered a hyperparameter.
+    num_random_walks : int
+        Number of random walks to try for each seed node.
+
+        Usually considered a hyperparameter.
+    num_neighbors : int
+        Number of neighbors to select for each seed.
+    weight_column : str, default "weights"
+        The weight of each neighbor, stored as an edge feature.
+
+    Inputs
+    ------
+    seed_nodes : Tensor
+        A tensor of seed node IDs of node type ``ntype``.
+
+    Outputs
+    -------
+    g : DGLHeteroGraph
+        A homogeneous graph constructed by selecting neighbors for each seed node according
+        to PinSAGE algorithm.
+
+    Examples
+    --------
+    See examples in :any:`PinSAGESampler`.
+    """
+    def __init__(self, G, ntype, metapath, random_walk_length, random_walk_restart_prob,
+                 num_random_walks, num_neighbors, weight_column='weights'):
+        self.G = G
+        self.ntype = ntype
+        self.weight_column = weight_column
+        self.num_random_walks = num_random_walks
+        self.num_neighbors = num_neighbors
+        self.random_walk_length = random_walk_length
+
+        self.metapath = metapath * random_walk_length
```

Comment:
We could let the `metapath` argument directly have repetition. Also, make sure to check the head and tail types are the same. Another question is about the restart. What if a walk restarts (halts) at a type different than the head type?

Reply:
See above.